scripts and ansible roles for gcp api server

ofersadan85 opened the pull request:
This will setup the host server for dev.encyclo-flower.com

~~2. It does not yet include swag settings / docker-compose.yml~~

Do not delete this branch after merging, it will continue to be developed

ofersadan85 commented:
@ohad24 I added documentation to the `README.md` on how to run the ansible commands

ofersadan85 commented:
@ohad24 This is now ready to be reviewed and merged, sorry for so many force-push, had a lot to cleanup

ohad24 commented on server/api.subfolder.conf:
```diff
@@ -0,0 +1,10 @@
+
+location ^~ /api {
```
I think I'll need this line `proxy_set_header Host $http_host;` also.
for the backend.
From https://www.uvicorn.org/deployment/#running-behind-nginx

ofersadan85 replied:
Will be fixed

ofersadan85 replied:
The line `include /config/nginx/proxy.conf` already includes that.
You can see what else is included: `docker exec swag cat /config/nginx/proxy.conf`

ohad24 replied:
Interesting

ohad24 commented on server/api.subfolder.conf:
```diff
@@ -0,0 +1,10 @@
+
+location ^~ /api {
```
Just make sure this is the correct path.
the API will serve from /api/v1/. It's already configure within the API code, So the location should be the same ?

ofersadan85 replied:
This will proxy everything under the API subfolder, including other sub folders such as v1

ohad24 replied:
Ohh, cool

ohad24 commented on server/docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+      - TZ=Asia/Jerusalem
+      - URL=encyclo-flower.com
+      - SUBDOMAINS=dev
+      - ONLY_SUBDOMAINS=true
```
This is necessary ?

ofersadan85 replied:
It is because the main domain points to a different server

ohad24 replied:
We'll need to remember to update this. maybe with a comment ?

ofersadan85 replied:
Open an issue

ofersadan85 replied:
Done in https://github.com/ohad24/encyclo-flower/pull/32/commits/fbdc3a0ce1df4b8dde80aaa0b8b72f930f5dfea9 But i still recommend to open an issue about it later

ohad24 replied:
Thanks.
Created new issue #34

ohad24 commented on server/docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+      - URL=encyclo-flower.com
+      - SUBDOMAINS=dev
+      - ONLY_SUBDOMAINS=true
+      - EMAIL=encyclo.flower@gmail.com
```
This is necessary ?

ofersadan85 replied:
This is optional, it's only for certificate notifications (but new certificates would be handled by SWAG anyway)

ohad24 replied:
cool

ohad24 commented on server/docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+      - SUBDOMAINS=dev
+      - ONLY_SUBDOMAINS=true
+      - EMAIL=encyclo.flower@gmail.com
+      - VALIDATION=http
```
For what purpose ?

ofersadan85 replied:
This is the certbot validation method, it must be here

ohad24 replied:
Maybe a small comment description ?

ofersadan85 replied:
Done in https://github.com/ohad24/encyclo-flower/pull/32/commits/fbdc3a0ce1df4b8dde80aaa0b8b72f930f5dfea9

ohad24 replied:
👌

ohad24 commented on server/docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+    networks:
+      - web
+
+  # api:
```
why marking out ?

ofersadan85 replied:
Until there is an image in the registry this will fail without commenting it out , will be fixed later

ohad24 replied:
OK, we'll get to this later

ohad24 commented on server/playbook.yml:
```diff
@@ -0,0 +1,82 @@
+    - name: Set hostname
+      hostname:
+        # NOTE: Manually change this to www for production
+        name: dev.encyclo-flower.com
```
Let's call the hostname something generic.. like server and a number
server01, no ?

ofersadan85 replied:
Right, will be fixed

ofersadan85 replied:
Done in https://github.com/ohad24/encyclo-flower/pull/32/commits/130a009d0db5e764b1f617ad7b260531065cbe3e

ohad24 replied:
👌

ohad24 commented on server/playbook.yml:
```diff
@@ -0,0 +1,82 @@
+        name: api
+        gid: 2000
+
+    - name: Create user for the api
```
I'm think we should use one user to the entire stack (webserver, FE and BE), no ?
maybe **ef** for encyclo-flower

ofersadan85 replied:
Right, will be fixed

ofersadan85 replied:
Done in https://github.com/ohad24/encyclo-flower/pull/32/commits/130a009d0db5e764b1f617ad7b260531065cbe3e

ohad24 replied:
👌

ohad24 commented on server/playbook.yml:
```diff
@@ -0,0 +1,82 @@
+        - docker-compose.yml
+        - api.subfolder.conf
+
+    - name: Run docker services
```
What if we want to run the stack with swarm ?

ohad24 replied:
or maybe not for now ?

ofersadan85 replied:
A whole lot of things will need to be changed for swarm / k8s... In that case we will probably not use Ansible (or we will use it only to prepare nodes)

ohad24 replied:
👌

ohad24 commented on README.md:
```diff
@@ -51,8 +51,13 @@ cd src/api
+    cd server
+    ansible-galaxy role install -r requirements.yml
+    ansible-galaxy collection install -r requirements.yml
+    # CREATE a 'hosts' file before the next step
```
Please add a small example for the hosts file to all the friends who never had experience with Ansible.
Thanks

ofersadan85 replied:
Right, will be fixed

ofersadan85 replied:
Done, see https://github.com/ohad24/encyclo-flower/pull/32/commits/37284ebe88a66ff7bccbb4aabe4db7d58ab34b1b

ohad24 replied:
🙏

ohad24 reviewed (commented):
Good work, Thanks.

I left a few topics and questions about the new files

ohad24 reviewed (approved):
![image](https://user-images.githubusercontent.com/33603349/160231839-756d5997-7693-4c5b-bc6e-c2c1b4b53712.png)

ofersadan85 commented:
Suggested edit:
```diff
diff --git a/.github/workflows/api.yml b/.github/workflows/api.yml
new file mode 100644
index 0000000..e0ef1f8
--- /dev/null
+++ b/.github/workflows/api.yml
@@ -0,0 +1,107 @@
+# This workflow should control anything to do with API
+# Including tests, building the docker image and pushing it
+# It should run on both development and production scenarios
+# Production workflow (tagging the image as "latest") should run only on tagged version push
+
+name: API workflow
+
+on:
+  workflow_dispatch:
+  push:
+    branches: [main]
+    tags:
+      - "v*.*.*"
+    paths:
+      # NOTE: Path filters are ignored for tag pushes
+      - "src/api/**"
+      - "tests/api/**"
+  pull_request:
+    branches: [main]
+    paths:
+      - "src/api/**"
+      - "tests/api/**"
+
+env:
+  DOCKER_IMAGE_NAME: ohad24/${{ github.event.repository.name }}-api
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install flake8 pytest
+          pip install -r src/api/requirements.txt
+      - name: Lint with flake8
+        run: |
+          # TODO: configure flake8
+          # stop the build if there are Python syntax errors or undefined names
+          flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+          # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+          flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+      - id: "auth"
+        name: "Authenticate to Google Cloud"
+        uses: "google-github-actions/auth@v0"
+        with:
+          credentials_json: "${{ secrets.GOOGLE_APPLICATION_CREDENTIALS }}"
+      - name: Test with pytest
+        env:
+          MONGO_URI: ${{ secrets.MONGO_URI }}
+        run: |
+          pytest tests/api/
+
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Metadata for Docker
+        # Docs: https://github.com/docker/metadata-action
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.DOCKER_IMAGE_NAME }}
+          flavor: |
+            latest=false
+          tags: |
+            type=ref,event=branch
+            type=ref,event=pr
+            type=semver,pattern={{major}},enable=${{ !startsWith(github.ref, 'refs/tags/v0.') }}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{version}}
+            type=edge
+            type=sha
+
+      - uses: actions/checkout@v2
+      - name: Login to DockerHub
+        # Docs: https://github.com/docker/login-action
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build and push
+        # Docs: https://github.com/docker/build-push-action
+        uses: docker/build-push-action@v2
+        with:
+          context: src/api
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+  deploy:
+    needs: [test, build] # This job will run only if other jobs were successful
+    runs-on: ubuntu-latest
+    steps:
+      - name: Trigger deploy via SSH
+        # Docs: https://github.com/appleboy/ssh-action
+        uses: appleboy/ssh-action@master
+        with:
+          host: ${{ secrets.SSH_DEPLOYMENT_HOST }}
+          port: ${{ secrets.SSH_DEPLOYMENT_PORT }}
+          username: ${{ secrets.SSH_DEPLOYMENT_USER }}
+          key: ${{ secrets.SSH_DEPLOYMENT_KEY }}
diff --git a/.github/workflows/docker-build-test.yml b/.github/workflows/docker-build-test.yml
deleted file mode 100644
index b308a3d..0000000
--- a/.github/workflows/docker-build-test.yml
+++ /dev/null
@@ -1,19 +0,0 @@
-name: API Docker Image build test
-
-on:
-  push:
-    branches: [ main ]
-    paths:
-      - 'src/api/**'
-  pull_request:
-    branches: [ main ]
-    paths:
-      - 'src/api/**'
-
-jobs:
-  build:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-    - name: Build the Docker image
-      run: docker build src/api --file src/api/Dockerfile --tag encyclo-flower-api-test:$(date +%s)
diff --git a/.github/workflows/docker-publish.yml b/.github/workflows/docker-publish.yml
deleted file mode 100644
index 48431a8..0000000
--- a/.github/workflows/docker-publish.yml
+++ /dev/null
@@ -1,49 +0,0 @@
-name: Publish Docker image
-
-on:
-  workflow_dispatch:
-  release:
-    types: [published]
-
-env:
-  DOCKER_IMAGE_NAME: ohad24/${{ github.event.repository.name }}
-
-jobs:
-  build_push_deploy:
-    name: Build & Push Docker image to Docker Hub
-    runs-on: ubuntu-latest
-
-    steps:
-      - name: Check out the repo
-        uses: actions/checkout@v2
-
-      - name: Build the Docker image
-        run: docker build src/api --file src/api/Dockerfile --tag ${DOCKER_IMAGE_NAME}
-
-      - name: Log in to Docker Hub
-        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
-        with:
-          username: ${{ secrets.DOCKER_USERNAME }}
-          password: ${{ secrets.DOCKER_PASSWORD }}
-      
-      - name: Extract metadata (tags, labels) for Docker
-        id: meta
-        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
-        with:
-          images: ${{ env.DOCKER_IMAGE_NAME }}
-      
-      - name: Build and push Docker image
-        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
-        with:
-          context: .
-          push: true
-          tags: ${{ steps.meta.outputs.tags }}
-          labels: ${{ steps.meta.outputs.labels }}
-
-      - name: Deploy via SSH
-        uses: appleboy/ssh-action@master
-        with:
-          host: ${{ secrets.SSH_DEPLOYMENT_HOST }}
-          port: ${{ secrets.SSH_DEPLOYMENT_PORT }}
-          username: ${{ secrets.SSH_DEPLOYMENT_USER }}
-          key: ${{ secrets.SSH_DEPLOYMENT_KEY }}
diff --git a/.github/workflows/python-tests.yml b/.github/workflows/python-tests.yml
deleted file mode 100644
index 2a0fbc1..0000000
--- a/.github/workflows/python-tests.yml
+++ /dev/null
@@ -1,51 +0,0 @@
-# This workflow will install Python dependencies, run tests and lint with a single version of Python
-# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
-
-name: API pytest
-
-on:
-  push:
-    branches: [ main ]
-    paths:
-      - 'src/api/**'
-      - 'tests/api/**'
-  pull_request:
-    branches: [ main ]
-    paths:
-      - 'src/api/**'
-      - 'tests/api/**'
-
-jobs:
-  build:
-
-    runs-on: ubuntu-latest
-
-    steps:
-    - uses: actions/checkout@v2
-    - name: Set up Python 3.10
-      uses: actions/setup-python@v2
-      with:
-        python-version: "3.10"
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install flake8 pytest
-        if [ -f src/api/requirements.txt ]; then pip install -r src/api/requirements.txt; fi
-    - name: Lint with flake8
-      run: |
-        # TODO: configure flake8
-        # stop the build if there are Python syntax errors or undefined names
-        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
-        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
-        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
-    - id: 'auth'
-      name: 'Authenticate to Google Cloud'
-      uses: 'google-github-actions/auth@v0'
-      with:
-        credentials_json: '${{ secrets.GOOGLE_APPLICATION_CREDENTIALS }}'
-    - name: Test with pytest
-      env:
-        MONGO_URI: ${{ secrets.MONGO_URI }}
-        CLOUD_BUCKET: ${{ secrets.CLOUD_BUCKET }}
-      run: |
-        pytest tests/api/

```
